packages/xo-server/src/xapi/mixins/vm.mjs

```diff
@@ -140,53 +143,99 @@ const methods = {
       hasBootableDisk = true
     }
 
-    // Modify existing (previous template) disks if necessary
-    existingVdis &&
-      (await Promise.all(
-        mapToArray(existingVdis, async ({ size, $SR: srId, ...properties }, userdevice) => {
-          const vbd = find(vm.$VBDs, { userdevice })
-          if (!vbd) {
+    // Compatibility with legacy code
+    if (existingVdis !== undefined) {
+      if (vdis === undefined) {
+        vdis = []
+      }
+
+      forEach(existingVdis, ({ $SR: sr, ...properties }, userdevice) => {
+        const vbd = find(vm.$VBDs, { userdevice })
+        if (!vbd) {
+          return
+        }
+
+        vdis.push({ ...properties, userdevice, sr })
+      })
+    }
+    // End compatibility with legacy code
+
+    if (vdis !== undefined && vdis.length > 0) {
+      const _vdisToCreate = []
+      const _vdisToUpdate = []
+      const _vdisToDestroy = []
+
+      vdis.forEach(vdi => {
+        if (vdi.userdevice === undefined) {
+          _vdisToCreate.push(vdi)
+          return
+        }
+
+        // If the userdevice match no vbd, create the VDI
+        const vbd = find(vm.$VBDs, { userdevice: vdi.userdevice })
+        if (vbd === undefined) {
+          if (vdi.destroy) {
+            log.warn('VDI ignored because it is marked as "destroy"', vdi)
             return
           }
-          let vdi = vbd.$VDI
-          await this._setObjectProperties(vdi, properties)
 
+          _vdisToCreate.push(vdi)
+          return
+        }
+
+        vdi.$ref = this.getObject(vbd.VDI).$ref
+        if (vdi.destroy) {
+          _vdisToDestroy.push(vdi)
+          return
+        }
+
+        _vdisToUpdate.push(vdi)
+      })
+
+      // TODO: set vm.suspend_SR
+      await Promise.all([
+        ..._vdisToDestroy.map(vdi => this.VDI_destroy(vdi.$ref)),
+        // Creates the user defined VDIs.
+        ..._vdisToCreate.map(async (vdi, i) => {
+          const vdiRef = await this.VDI_create({
+            name_description: vdi.name_description,
+            name_label: vdi.name_label,
+            virtual_size: vdi.size,
+            SR: this.getObject(vdi.sr, 'SR').$ref,
+          })
+          $defer.onFailure(() => this.VDI_destroy(vdiRef))
+
+          await this.VBD_create({
+            // Either the CD or the 1st disk is bootable (only useful for PV VMs)
+            bootable: !(hasBootableDisk || i),
```

Comment:
I can't leave a comment on the right line, but I think hasBootbaleDisk should  be (re) computed after destroying the vdi ( the only bootable disk may have been destroyed)